# Conceitos Básicos de Python

Visualizar detalhes e parâmetros de uma função (**shift** + **tab**)

## Indentação e Definições

##### Organização do Código (Indentação)

In [ ]:
x = 1
if x == 1:
  print('Você está vendo que tudo que está abaixo de : tem que ser indentado')

Você está vendo que tudo que está abaixo de : tem que ser indentado


In [ ]:
if x == 1:
print('Você está vendo que tudo que está abaixo de : tem que ser indentado') # Erro de Indentação

Você está vendo que tudo que está abaixo de : tem que ser indentado


In [ ]:
# definições das funções

def funcao_a():
    print('A) Este é outro exemplo de indentação necessária')
print("=> esta string está fora da função e vai imprimir no momento de *definição*")

def funcao_b():
    print('B) Este é outro exemplo de indentação necessária')
    print("   => esta string está dentro da função B e vai imprimir no momento da *chamada*")

=> esta string está fora da função e vai imprimir no momento de *definição*


In [ ]:
# chamadas das funções

funcao_a()
funcao_b()

A) Este é outro exemplo de indentação necessária
B) Este é outro exemplo de indentação necessária
   => esta string está dentro da função B e vai imprimir no momento da *chamada*


## Variáveis e Tipos

### `int`: números inteiros

In [ ]:
myint = 5
print(myint)
print(type(myint))

5
<class 'int'>


### `float`: números com parte fracionária

In [ ]:
myfloat = 5.1
print(myfloat)
print(type(myfloat))

5.1
<class 'float'>


### `str`: strings, ou "sequências de caracteres de texto"

In [ ]:
mystring = 'Isso é uma string' # aspas duplas também funcionam
print(mystring)
print(type(mystring))

Isso é uma string
<class 'str'>


### `bool`: booleano ou binário

In [ ]:
print(True, 'equivale a', int(True))
print(False, 'equivale a', int(False))
print(type(True))

True equivale a 1
False equivale a 0
<class 'bool'>


### `None`: informação faltante

In [ ]:
list01 = [0, 'x', 3, None, 'b']

count = 1
for i in list01:
    if i is None:
        print(f"\nO {count}º elemento é uma informação faltante\n")
    else:
        print(f"O {count}º elemento é igual a {i}")
    count += 1

O 1º elemento é igual a 0
O 2º elemento é igual a x
O 3º elemento é igual a 3

O 4º elemento é uma informação faltante

O 5º elemento é igual a b


## Operações

##### Matemática

In [ ]:
a = 5
b = 2

print(f'a = {a} \nb = {b}\n')
print('Soma: a + b =', a + b)
print('Subtração: a - b =', a - b)
print('Multiplicação: a * b =', a * b)
print('Divisão: a / b =', a / b)
print('Potenciação: a ** b =', a ** b)
print('Resto: a % b =', a % b)
print('Divisão inteira: a // b =', a // b)

a = 5 
b = 2

Soma: a + b = 7
Subtração: a - b = 3
Multiplicação: a * b = 10
Divisão: a / b = 2.5
Potenciação: a ** b = 25
Resto: a % b = 1
Divisão inteira: a // b = 2


In [ ]:
a = 'Já sou'
b = 'expert'
c = 'em Python?'

print(a + b + c)
print(a + ' ' + b + ' ' + c)
print(b * 10)

Já souexpertem Python?
Já sou expert em Python?
expertexpertexpertexpertexpertexpertexpertexpertexpertexpert


In [ ]:
print(a - b) # descomente essa linha e conclua que strings não aceitam operações de subtração

TypeError: ignored

In [ ]:
a = 9
print(a)
a += 1
print(a)
a -= 1
a = a - 1
print(a)
a /= 2
a = a / 2
print(a)
a *= 2
a = a * 2
print(a)

9
10
8
2.0
8.0


##### Comparações

In [ ]:
a = 5
b = 2

print(f'a = {a} \nb = {b}\n')
print('Igual: a == 5\n' + str(a == 5))
print('Diferente: a != b\n', a != b)
print('Maior: a > b\n', a > b)
print('Maior ou Igual: a >= b\n', a >= b)
print('Menor: a < b\n', a < b)
print('Menor ou Igual: a <= b\n', a <= b)

a = 5 
b = 2

Igual: a == 5
True
Diferente: a != b
 True
Maior: a > b
 True
Maior ou Igual: a >= b
 True
Menor: a < b
 False
Menor ou Igual: a <= b
 False


In [ ]:
print("a = 5") # constate que há erro de sintaxe aqui... o que está errado?

a = 5


Não confundir atribuição `a = b` _("a recebe b")_ com verificação de igualdade `a == b` _("a é igual a b?")_

##### Lógica

In [ ]:
a = 1
b = 2
c = 1

print('a < b:', a < b)
print('c == a:', c == a)

a < b: True
c == a: True


AND

In [ ]:
print('a == b:', a == b)
print('c == a:', c == a)
print('(a == b) and (c == a):', (a == b) and (c == a))

a == b: False
c == a: True
(a == b) and (c == a): False


In [ ]:
print(True and True)
print(False and True)
print(True and False)
print(False and False)

True
False
False
False


OR

In [ ]:
print('a < b:', a < b)
print('c > a:', c > a)
print('(a < b) or (c > a):', (a < b) or (c > a))

a < b: True
c > a: False
(a < b) or (c > a): True


In [ ]:
print(True or True)
print(False or True)
print(True or False)
print(False or False)

True
True
True
False


NOT

In [ ]:
a = 10
print('a == 10:', a == 10)
print('not(a == 10):', not(a == 10))

a == 10: True
not(a == 10): True


# Conceitos Básicos de Python

Visualizar detalhes e parâmetros de uma função (**shift** + **tab**)

## Estruturas de Dados _(built-in)_

Documentação - https://docs.python.org/3/library/functions.html

### `list`: listas são sequências mutáveis

##### Inicialização

In [ ]:
mylist = []
print(type(mylist))

<class 'list'>


##### Adição de elementos

In [ ]:
mylist.append(10)
print(mylist)

# Teste rodar essa célula mais de uma vez... o que acontece?

[8, 8, 8, 10]


`append()` é um método _inplace_.

Dizemos que um método é _inplace_ quando ela modifica o objeto envolvido. No caso de fazermos `x.append(item)`, a lista `x` é alterada durante a execução da função, passando a incluir o elemento `item`.

##### Adição de elementos em posição específica

In [ ]:
mylist2 = ['Ana', 9, '35', 3.6, 0]

#inserindo elemento entre o terceiro e quarto elemento
mylist2.insert(3 , 10)
mylist2

['Ana', 9, '35', 10, 3.6, 0]

##### Indexação

In [ ]:
mylist = ['Alves, R.',
 'Balena, A.',
 'Carvalho, F.',
 'Coelho, H.',
 'Cristino, L.',
 'Donzelli, S.',
 'Freire, D.',
]

In [ ]:
# Como seleciono itens específicos dessa lista?
print(mylist[0])
print(mylist[1])
print(mylist[-1])
print(mylist[-2])

Alves, R.
Balena, A.
Freire, D.
Donzelli, S.


In [ ]:
print(mylist[30]) # Erro de indexação - o que aconteceu?

IndexError: ignored

In [ ]:
# E se quisermos percorrer cada elemento da lista? (veremos em mais detalhes depois)
for item in mylist:
    print(item)

Alves, R.
Balena, A.
Carvalho, F.
Coelho, H.
Cristino, L.
Donzelli, S.
Freire, D.


##### Operações adicionais

Soma de listas

In [ ]:
# Vamos criar uma nova lista que aponta (é igual) a mylist
newlist = mylist
print(newlist)

['Alves, R.', 'Balena, A.', 'Carvalho, F.', 'Coelho, H.', 'Cristino, L.', 'Donzelli, S.', 'Freire, D.']


In [ ]:
print(mylist + newlist)

['Alves, R.', 'Balena, A.', 'Carvalho, F.', 'Coelho, H.', 'Cristino, L.', 'Donzelli, S.', 'Freire, D.', 'Alves, R.', 'Balena, A.', 'Carvalho, F.', 'Coelho, H.', 'Cristino, L.', 'Donzelli, S.', 'Freire, D.']


In [ ]:
print(3 * mylist)

['Alves, R.', 'Balena, A.', 'Carvalho, F.', 'Coelho, H.', 'Cristino, L.', 'Donzelli, S.', 'Freire, D.', 'Alves, R.', 'Balena, A.', 'Carvalho, F.', 'Coelho, H.', 'Cristino, L.', 'Donzelli, S.', 'Freire, D.', 'Alves, R.', 'Balena, A.', 'Carvalho, F.', 'Coelho, H.', 'Cristino, L.', 'Donzelli, S.', 'Freire, D.']


Tamanho

In [ ]:
len(mylist)

7

In [ ]:
mylist[6] == mylist[-1]

True

Índice de um elemento

In [ ]:
mylist.index('Alves, R.')

0

In [ ]:
mylist[0]

'Alves, R.'

Número de ocorrências

In [ ]:
mylist

['Alves, R.',
 'Balena, A.',
 'Carvalho, F.',
 'Coelho, H.',
 'Cristino, L.',
 'Donzelli, S.',
 'Freire, D.']

In [ ]:
print('Quantas ocorrências em mylist tem "Alves, R."?\n', mylist.count('Alves, R.'))
print('Quantas ocorrências em (mylist * 3) tem "Alves, R."?\n', (mylist * 3).count('Alves, R.'))

Quantas ocorrências em mylist tem "Alves, R."?
 1
Quantas ocorrências em (mylist * 3) tem "Alves, R."?
 3


Está contido

In [ ]:
print('Alves, R. está em mylist?\n', 'Alves, R.' in mylist)
print('Sugimoto, P. está em mylist?\n', 'Sugimoto, P.' in mylist)

Alves, R. está em mylist?
 True
Sugimoto, P. está em mylist?
 False


Slices

In [ ]:
n = 10
lista = list(range(n))
lista

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [ ]:
lista[n-1]

IndexError: ignored

In [ ]:
print(lista[:]) # todos os elementos
print(lista[:3]) # os 3 primeiros
print(lista[-2:]) # dois últimos
print(lista[1:5]) # do segundo ao quinto elemento
print(lista[1:5:2]) # do segundo ao quinto elemento com passo 2
print(lista[::-1]) # forma fácil de inverter a lista

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
[0, 1, 2]
[8, 9]
[1, 2, 3, 4]
[1, 3]
[9, 8, 7, 6, 5, 4, 3, 2, 1, 0]


Multiplicação por constante

In [ ]:
a = [1, 2, 3, 4]
b = [a, a]

print(a * 2)
print(b * 2)

[1, 2, 3, 4, 1, 2, 3, 4]
[[1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4], [1, 2, 3, 4]]


In [ ]:
b[0]

[1, 2, 3, 4]

Reatribuição de valores

In [ ]:
a = [1, 2, 3, 4]
print(a)
a[0] = "x"
a[-1] = 234562873568236
print(a)

[1, 2, 3, 4]
['x', 2, 3, 234562873568236]


Remoção de elementos

In [ ]:
print(mylist)
mylist.pop(0)
print()
print(mylist)

['Balena, A.', 'Carvalho, F.', 'Coelho, H.', 'Cristino, L.', 'Donzelli, S.', 'Freire, D.']

['Carvalho, F.', 'Coelho, H.', 'Cristino, L.', 'Donzelli, S.', 'Freire, D.']


Teste rodar a célula anterior mais de uma vez... o que acontece?

`pop()`, assim como `append()`, também é uma operação _inplace_.

In [ ]:
x = list(enumerate(mylist))

[(0, 'Carvalho, F.'),
 (1, 'Coelho, H.'),
 (2, 'Cristino, L.'),
 (3, 'Donzelli, S.'),
 (4, 'Freire, D.')]

#### Exercício - listas

**Ex. 1)** Retorne o segundo elemento, dentro do segundo elemento de `ex_list`

In [ ]:
ex_list = [['a', 1], [True, 3 ** 2]]
ex_list

[['a', 1], [True, 9]]

In [ ]:
# Seu código aqui
ex_list[1][1]

9

### `set`: conjuntos são coleções não-ordenadas de elementos

Inicialização

In [ ]:
a = [1, 2, 3]
b = set(a)

In [ ]:
a

[1, 2, 3]

In [ ]:
b

{1, 2, 3}

In [ ]:
a = {1, 2, 3}
type(a)

set

In [ ]:
myset1 = set([1, 1, 'Minha String', 4, True])
myset2 = set(['Minha String', 10])

In [ ]:
print(myset1)
print(myset2)

{1, 4, 'Minha String'}
{10, 'Minha String'}


Operações entre conjuntos (ver mais em  https://docs.python.org/3/library/sets.html)

In [ ]:
# Union
myset1.union(myset2)
myset1 | myset2

{1, 10, 4, 'Minha String'}

In [ ]:
# Intersecção
print(myset1.intersection(myset2))
print(myset1 & myset2)

{'Minha String'}
{'Minha String'}


### `dict`: dicionários são mapeamentos _chave_ --> _valor_

##### Inicialização

In [ ]:
mydict = {}
mydict

{}

In [ ]:
# É necessário que a chave seja um elemento não mutável
# se houver duas ou mais chaves de mesmo valor, apenas uma é considerada

mydict = {'key1': 1,
          'chave2': [2., 3, 'a'],
          'key3': 'valor',
          'key3': 0,
          45: 'x',
         (2,3,5): 3}

mydict

{'key1': 1, 'chave2': [2.0, 3, 'a'], 'key3': 0, 45: 'x', (2, 3, 5): 3}

##### Indexação

In [ ]:
mydict

{'key1': 1, 'chave2': [2.0, 3, 'a'], 'key3': 0, 45: 'x', (2, 3, 5): 3}

In [ ]:
mydict['key3']

0

##### Adição/Remoção de Elementos

In [ ]:
mydict['key4'] = [1, 2, 3]
mydict['key5'] = mydict['key4']
mydict

{'key1': 1,
 'chave2': [2.0, 3, 'a'],
 'key3': 0,
 45: 'x',
 (2, 3, 5): 3,
 'key4': [1, 2, 3],
 'key5': [1, 2, 3]}

In [ ]:
mydict['key4'].append("x")
mydict['key5']

[1, 2, 3, 'x']

In [ ]:
mydict

{'key1': 1,
 'chave2': [2.0, 3, 'a'],
 'key3': 0,
 45: 'x',
 (2, 3, 5): 3,
 'key4': [1, 2, 3, 'x'],
 'key5': [1, 2, 3, 'x']}

In [ ]:
print(mydict.pop('key4'))
print(mydict)

[1, 2, 3, 'x']
{'key1': 1, 'chave2': [2.0, 3, 'a'], 'key3': 0, 45: 'x', (2, 3, 5): 3, 'key5': [1, 2, 3, 'x']}


###### Operações Adicionais

Listando itens

In [ ]:
list(mydict.items())

[('key1', 1),
 ('chave2', [2.0, 3, 'a']),
 ('key3', 0),
 (45, 'x'),
 ((2, 3, 5), 3),
 ('key5', [1, 2, 3, 'x'])]

Listando keys

In [ ]:
list(mydict.keys())

['key1', 'chave2', 'key3', 45, (2, 3, 5), 'key5']

Listando values

In [ ]:
list(mydict.values())

[1, [2.0, 3, 'a'], 0, 'x', 3, [1, 2, 3, 'x']]

Atualização de valores

In [ ]:
# Alternativa 1: redefinição
print(mydict)
mydict['key1'] = 50
print(mydict)

{'key1': 1, 'chave2': [2.0, 3, 'a'], 'key3': 0, 45: 'x', (2, 3, 5): 3, 'key5': [1, 2, 3, 'x']}
{'key1': 50, 'chave2': [2.0, 3, 'a'], 'key3': 0, 45: 'x', (2, 3, 5): 3, 'key5': [1, 2, 3, 'x']}


In [ ]:
# Alternativa 2: update
print(mydict)
mydict.update({'key1': 10, 'key2': 20}) # operação Inplace
print(mydict)

{'key1': 50, 'chave2': [2.0, 3, 'a'], 'key3': 0, 45: 'x', (2, 3, 5): 3, 'key5': [1, 2, 3, 'x']}
{'key1': 10, 'chave2': [2.0, 3, 'a'], 'key3': 0, 45: 'x', (2, 3, 5): 3, 'key5': [1, 2, 3, 'x'], 'key2': 20}


#### Exercício - dicionários

**Ex. 2)** Mude a idade de `Alves` para 30

In [ ]:
ex_dict = {
     'pessoa1': {
         'nome': 'Alves',
         'idade': 25
     },
     'pessoa2': {
         'nome': 'Martins',
         'idade': 30
     }
}
ex_dict

{'pessoa1': {'nome': 'Alves', 'idade': 25},
 'pessoa2': {'nome': 'Martins', 'idade': 30}}

In [ ]:
# Seu código aqui
ex_dict['pessoa1']['idade'] = 30
ex_dict

{'pessoa1': {'nome': 'Alves', 'idade': 30},
 'pessoa2': {'nome': 'Martins', 'idade': 30}}

### `str`: strings são cadeias de caracteres (texto)

In [ ]:
mystr = 'Analytics Skills Accelerator'
mystr = "Analytics Skills Accelerator"

print(mystr[0])
print(mystr[:9])
print(mystr[-11:])

A
Analytics
Accelerator


In [ ]:
for index, char in enumerate(mystr):
    print(index, char)

0 A
1 n
2 a
3 l
4 y
5 t
6 i
7 c
8 s
9  
10 S
11 k
12 i
13 l
14 l
15 s
16  
17 A
18 c
19 c
20 e
21 l
22 e
23 r
24 a
25 t
26 o
27 r


In [ ]:
mystr

'Analytics Skills Accelerator'

In [ ]:
mystr.split(' ')

['Analytics', 'Skills', 'Accelerator']

In [ ]:
print(mystr)
print(mystr.upper()) # Upper case
print(mystr.lower()) # Lower case

Analytics Skills Accelerator
ANALYTICS SKILLS ACCELERATOR
analytics skills accelerator


In [ ]:
mystr = mystr.replace('-', '') # Substituição
print(mystr)

AnalyticsSkillsAccelerator


##### f-string (https://fstring.help/)

In [ ]:
var1 = 1
var2 = 'a'
var3 = 0.6584525
print(f'var1 is equal to {var1}')
print(f'var2 is equal to {var2}')
print(f'var3 is equal to {var3:.2%}')

var1 is equal to 1
var2 is equal to a
var3 is equal to 65.85%


#### Exercício - strings

**Ex. 3)** Encontre a **sexta letra da segunda palavra** da string abaixo e retorne a letra **em maiúsculo**

In [ ]:
mystring = 'Analytics Skills Accelerator'

In [ ]:
# Seu código aqui

## Fluxo de execução

### Condicionais

##### IF, ELIF e ELSE

In [ ]:
a = 10 # altere esse valor de modo a entrar nos diferentes casos de condicional
if a < 10:
    print('Valor menor que 10')
elif a > 10:
    print('Valor maior que 10')
else:
    print('Valor igual a 10')

Valor igual a 10


In [ ]:
a = 10 # altere esse valor de modo a entrar nos diferentes casos de condicional
if (a == 10) and (type(a) == int):
    print('Valor inteiro igual a 10')
elif (a == 10) and (type(a) == float):
    print('Valor float igual a 10')
elif (a == '10') and (type(a) == str):
    print('String com número 10')
else:
    print('-')
print('---')

Valor inteiro igual a 10
---


In [ ]:
a = 10. # Mostrando a diferença entre if e elif
if a == 10:
    print('Valor igual a 10')
elif a == 10.0: # Mude essa linha para if
    print('Valor float igual a 10')
else:
    print('-')

Valor igual a 10


In [ ]:
# Vamos definir novamente a lista de nomes
mylist = ['Alves, R.',
 'Balena, A.',
 'Carvalho, F.',
 'Coelho, H.',
 'Cristino, L.',
 'Donzelli, S.',
 'Freire, D.',
]

In [ ]:
nome = 'Takehara, D.' # altere esse valor de modo a entrar nos diferentes casos de condicional
if nome in mylist:
    print('%s está na lista de nomes!' % nome)
else:
    print('%s não está na lista de nomes!' % nome)

Takehara, D. não está na lista de nomes!


In [ ]:
a = 10
answer = 'Igual a 10' if (a == 10) else 'Diferente de 10'
print(answer)

Igual a 10


### Loops

##### FOR

In [ ]:
iterador1 = [1, 2, 3, 4, 5] # lista
iterador2 = (1, 2, 3, 4, 5) # tupla
iterador3 = range(1, 6) # método range
iterador4 = {'key1': 1, 'key2': 2, 'key3': 3} # dicionário

In [ ]:
# lista
for item in iterador1:
    print(item * 2)

2
4
6
8
10


In [ ]:
# tupla
for i in iterador2:
    print(i)

1
2
3
4
5


In [ ]:
# método range
for i in iterador3:
    print(i)

1
2
3
4
5


In [ ]:
#dicionário
iterador4

{'key1': 1, 'key2': 2, 'key3': 3}

In [ ]:
# O que acontece se iteramos sobre um dicionário?
for i in iterador4:
    print(i)

key1
key2
key3


In [ ]:
# E se quisermos recuperar os valores do dicionário também?
for i in iterador4:
    print(i, iterador4[i])

key1 1
key2 2
key3 3


In [ ]:
for key, value in iterador4.items():
    print(key, value)

key1 1
key2 2
key3 3


##### WHILE

In [ ]:
# # O que aconterá se rodarmos essa célula?
# count = 0
# while count < 5:
#     print(count)

In [ ]:
# O que mudou?
count = 0
while count < 5:
    print(count)
    count += 1

0
1
2
3
4


##### BREAK e CONTINUE

In [ ]:
count = 0
while True:
    print(count)
    count += 1
    if count == 8:
        break

0
1
2
3
4
5
6
7


In [ ]:
# todo código após o continue é ignorado, ele retorna para primeiro comando dentro do looping
# comente o 'continue' e veja o que ocorre
for i in range(10):
    if i % 2 == 0:
        print(f'{i} é par')
        continue
    print(f'{i} é impar')

0 é par
1 é impar
2 é par
3 é impar
4 é par
5 é impar
6 é par
7 é impar
8 é par
9 é impar


Há vários sites com sessões introdutórias como essas. Tente acessar https://www.w3schools.com/python/ ou https://www.datacamp.com/courses/intro-to-python-for-data-science

### Funções

In [ ]:
# exemplo de função sem nenhum argumento
def funcao_sem_argumentos():
    print("* funcao_sem_argumentos")

In [ ]:
funcao_sem_argumentos()

* funcao_sem_argumentos


In [ ]:
# exemplo de função com um argumento
def funcao_com_um_argumento(x):
    print(f"* funcao_com_um_argumento (x = {x} ({type(x)}))")

In [ ]:
funcao_com_um_argumento("a")
for i in range(5):
    funcao_com_um_argumento(i)

* funcao_com_um_argumento (x = a (<class 'str'>))
* funcao_com_um_argumento (x = 0 (<class 'int'>))
* funcao_com_um_argumento (x = 1 (<class 'int'>))
* funcao_com_um_argumento (x = 2 (<class 'int'>))
* funcao_com_um_argumento (x = 3 (<class 'int'>))
* funcao_com_um_argumento (x = 4 (<class 'int'>))


In [ ]:
# função com argumento opcional

def media(x, y, tipo="aritmetica"):
    """
    Calcula a média de um par de números.

    Parameters
    ----------
    x : int or float
        Primeiro número.
    y : int or float
        Segundo número.
    tipo : str, default 'aritmetica'
        Tipo de média a ser calculado, que pode ser 'aritmetica' ou 'geometrica'.

    Returns
    -------
    float
        A média entre ``x`` e ``y``.
    """
    # acima, vemos o exemplo de uma doc-string, que documenta a função

    if tipo == "aritmetica":
        return (x + y) / 2
    elif tipo == "geometrica":
        return (x * y) ** 0.5

In [ ]:
# por padrão, o tipo considerado é "aritmetica"
media(1, 2)

1.5

In [ ]:
# isso pode ser passado explicitamente, e o resultado é o mesmo
media(1, 2, tipo="aritmetica")

1.5

In [ ]:
# apenas se o tipo for alterado para "geometrica" o resultado muda
media(1, 2, tipo="geometrica")

1.4142135623730951

### Exercícios - Condicionais e Loops

**Ex. 4)** Itere na lista **numbers** e imprima todos os números ímpares na mesma ordem que eles aparecem na lista. Além disso, não imprima números que vêm depois do 237 na sequência.

In [4]:
numbers = [
    951, 402, 984, 651, 360, 69, 408, 319, 601, 485, 980, 507, 725, 547, 544,
    615, 83, 165, 141, 501, 263, 617, 865, 575, 219, 390, 984, 592, 236, 105, 942, 941,
    386, 462, 47, 418, 907, 344, 236, 375, 823, 566, 597, 978, 328, 615, 953, 345,
    399, 162, 758, 219, 918, 237, 412, 566, 826, 248, 866, 950, 626, 949, 687, 217,
    815, 67, 104, 58, 512, 24, 892, 894, 767, 553, 81, 379, 843, 831, 445, 742, 717,
    958, 609, 842, 451, 688, 753, 854, 685, 93, 857, 440, 380, 126, 721, 328, 753, 470,
    743, 527
]

In [5]:
# Seu código aqui
for n in numbers:
    if n%2 != 0:
        print(f'{n} é um número impar da lista')
    if n == 237:
        break

951 é um número impar da lista
651 é um número impar da lista
69 é um número impar da lista
319 é um número impar da lista
601 é um número impar da lista
485 é um número impar da lista
507 é um número impar da lista
725 é um número impar da lista
547 é um número impar da lista
615 é um número impar da lista
83 é um número impar da lista
165 é um número impar da lista
141 é um número impar da lista
501 é um número impar da lista
263 é um número impar da lista
617 é um número impar da lista
865 é um número impar da lista
575 é um número impar da lista
219 é um número impar da lista
105 é um número impar da lista
941 é um número impar da lista
47 é um número impar da lista
907 é um número impar da lista
375 é um número impar da lista
823 é um número impar da lista
597 é um número impar da lista
615 é um número impar da lista
953 é um número impar da lista
345 é um número impar da lista
399 é um número impar da lista
219 é um número impar da lista
237 é um número impar da lista


**Ex. 5)** Vasculhando a base de faturamento de uma empresa fictícia, encontramos a lista de SKUs mais vendidos (**top_skus**). Contudo, não precisamos de todas as informações descritas, apenas da **família (primeiras 3 letras)** e o **formato**, conforme o exemplo abaixo:

`VEVQ340 Z Y BR 0490X0550 AB 00300 10 099` -> `VEV_0490X0550`

Dados os materiais  na lista **top_skus**, crie uma segunda lista (**top_skus_short**), abreviados pela regra de família e formato.

In [6]:
top_skus = [
    'VEVQ340 Z Y BR 0490X0550 AB 00300 10 099',
    'VEVY680 Z Y BR 0470X1140 AB 00150 02 099',
    'KVNX880 Z Y BR 0190X0840 AB 00250 02 089',
    'VEVT190 Z Y BR 0340X0660 AB 00450 05 099',
    'UTSO140 Z Y BR 0400X0470 AB 00150 02 089',
    'KVNL330 Z Y BR 0500X1040 AB 00300 05 089',
    'WTEQ760 Z Y BR 0230X0230 AB 00450 02 089',
    'SLUL970 Z Y BR 0450X0770 AB 00150 02 089',
    'SLUY170 Z Y BR 0280X0870 AB 00500 02 099',
    'VEVG170 Z Y BR 0310X0900 AB 00250 05 099',
    'VEVH720 Z Y BR 0410X0800 AB 00500 02 099',
    'VEVN770 Z Y BR 0150X0270 AB 00200 02 099',
    'SLUQ150 Z Y BR 0330X0980 AB 00200 05 089',
    'SLUB390 Z Y BR 0420X0530 AB 00500 02 099',
    'KVNC610 Z Y BR 0120X0690 AB 00200 02 089',
    'SLUM990 Z Y BR 0220X0690 AB 00350 02 099',
    'VEVO290 Z Y BR 0400X1170 AB 00150 02 099',
    'VEVV140 Z Y BR 0330X0630 AB 00150 02 099',
    'SLUL400 Z Y BR 0270X0910 AB 00350 02 099',
    'KVNM600 Z Y BR 0210X0770 AB 00200 05 089',
    'VEVO820 Z Y BR 0470X0790 AB 00450 02 099',
    'SLUI380 Z Y BR 0460X0970 AB 00450 10 089',
    'SLUL810 Z Y BR 0110X0720 AB 00200 05 089',
    'KVNL210 Z Y BR 0120X1180 AB 00200 02 089',
    'SLUL630 Z Y BR 0400X1130 AB 00250 02 089',
    'KNIJ590 Z Y BR 0300X0780 AB 00450 02 089',
    'VEVP760 Z Y BR 0320X0880 AB 00350 02 099',
    'KVNH160 Z Y BR 0450X0610 AB 00300 02 089',
    'VEVT190 Z Y BR 0170X1080 AB 00150 02 099',
    'VEVK200 Z Y BR 0260X1050 AB 00300 02 099',
    'XTHL520 Z Y BR 0330X0820 AB 00150 02 089',
    'KVNW920 Z Y BR 0120X0450 AB 00400 02 089',
    'SLUU930 Z Y BR 0320X1090 AB 00250 02 089',
    'RTBC830 Z Y BR 0480X0550 AB 00250 10 089',
]

In [7]:
# Seu código aqui
#Solução 1
top_skus_short = []

for sku in top_skus:
    top_skus_short.append(f'{sku[:3]}_{sku.split()[4]}')

top_skus_short

['VEV_0490X0550',
 'VEV_0470X1140',
 'KVN_0190X0840',
 'VEV_0340X0660',
 'UTS_0400X0470',
 'KVN_0500X1040',
 'WTE_0230X0230',
 'SLU_0450X0770',
 'SLU_0280X0870',
 'VEV_0310X0900',
 'VEV_0410X0800',
 'VEV_0150X0270',
 'SLU_0330X0980',
 'SLU_0420X0530',
 'KVN_0120X0690',
 'SLU_0220X0690',
 'VEV_0400X1170',
 'VEV_0330X0630',
 'SLU_0270X0910',
 'KVN_0210X0770',
 'VEV_0470X0790',
 'SLU_0460X0970',
 'SLU_0110X0720',
 'KVN_0120X1180',
 'SLU_0400X1130',
 'KNI_0300X0780',
 'VEV_0320X0880',
 'KVN_0450X0610',
 'VEV_0170X1080',
 'VEV_0260X1050',
 'XTH_0330X0820',
 'KVN_0120X0450',
 'SLU_0320X1090',
 'RTB_0480X0550']

In [8]:
# Seu código aqui
#Solução2
top_skus_short = []

for sku in top_skus:
    top_skus_short.append(
        sku[:3]
        + '_'
        + sku.split()[4]
    )

top_skus_short

['VEV_0490X0550',
 'VEV_0470X1140',
 'KVN_0190X0840',
 'VEV_0340X0660',
 'UTS_0400X0470',
 'KVN_0500X1040',
 'WTE_0230X0230',
 'SLU_0450X0770',
 'SLU_0280X0870',
 'VEV_0310X0900',
 'VEV_0410X0800',
 'VEV_0150X0270',
 'SLU_0330X0980',
 'SLU_0420X0530',
 'KVN_0120X0690',
 'SLU_0220X0690',
 'VEV_0400X1170',
 'VEV_0330X0630',
 'SLU_0270X0910',
 'KVN_0210X0770',
 'VEV_0470X0790',
 'SLU_0460X0970',
 'SLU_0110X0720',
 'KVN_0120X1180',
 'SLU_0400X1130',
 'KNI_0300X0780',
 'VEV_0320X0880',
 'KVN_0450X0610',
 'VEV_0170X1080',
 'VEV_0260X1050',
 'XTH_0330X0820',
 'KVN_0120X0450',
 'SLU_0320X1090',
 'RTB_0480X0550']

**Ex. 6)** Agrupe os SKUs da lista **top_skus** naqueles de mesma família. Armazene a informação em um dicionário cujas chave é uma família e o valor é respectiva lista de SKUs. Por exemplo:

    top_skus = [
        'VEVQ340 Z Y BR 0490X0550 AB 00300 10 099',
        'VEVY680 Z Y BR 0470X1140 AB 00150 02 099',
        'KVNX880 Z Y BR 0190X0840 AB 00250 02 089',
    ]

Seu output deverá ser algo do como:

    dict_skus = {
        'KVN': ['KVNX880 Z Y BR 0190X0840 AB 00250 02 089'],
        'VEV': [
            'VEVQ340 Z Y BR 0490X0550 AB 00300 10 099',
            'VEVY680 Z Y BR 0470X1140 AB 00150 02 099',
        ],
    }

In [9]:
# Seu código aqui
dict_skus = {}

for sku in top_skus:
    familia = sku[:3]
    if familia in dict_skus.keys(): #se a familia já existe
        dict_skus[familia].append(sku) #adiciono o elemento a lista
    else: #caso não exista
        dict_skus[familia] = [sku] #adiciono a nova chave da familia e respectivo valor do sku

dict(sorted(dict_skus.items())) #visualizando o dicionário em ordem alfabética

{'KNI': ['KNIJ590 Z Y BR 0300X0780 AB 00450 02 089'],
 'KVN': ['KVNX880 Z Y BR 0190X0840 AB 00250 02 089',
  'KVNL330 Z Y BR 0500X1040 AB 00300 05 089',
  'KVNC610 Z Y BR 0120X0690 AB 00200 02 089',
  'KVNM600 Z Y BR 0210X0770 AB 00200 05 089',
  'KVNL210 Z Y BR 0120X1180 AB 00200 02 089',
  'KVNH160 Z Y BR 0450X0610 AB 00300 02 089',
  'KVNW920 Z Y BR 0120X0450 AB 00400 02 089'],
 'RTB': ['RTBC830 Z Y BR 0480X0550 AB 00250 10 089'],
 'SLU': ['SLUL970 Z Y BR 0450X0770 AB 00150 02 089',
  'SLUY170 Z Y BR 0280X0870 AB 00500 02 099',
  'SLUQ150 Z Y BR 0330X0980 AB 00200 05 089',
  'SLUB390 Z Y BR 0420X0530 AB 00500 02 099',
  'SLUM990 Z Y BR 0220X0690 AB 00350 02 099',
  'SLUL400 Z Y BR 0270X0910 AB 00350 02 099',
  'SLUI380 Z Y BR 0460X0970 AB 00450 10 089',
  'SLUL810 Z Y BR 0110X0720 AB 00200 05 089',
  'SLUL630 Z Y BR 0400X1130 AB 00250 02 089',
  'SLUU930 Z Y BR 0320X1090 AB 00250 02 089'],
 'UTS': ['UTSO140 Z Y BR 0400X0470 AB 00150 02 089'],
 'VEV': ['VEVQ340 Z Y BR 0490X0550 AB 00

# Quiz Time

## Quiz Q.2.1

Você usará a lista `top_skus_count` para completar o quiz. Essa é uma lista
contendo tuplas (pares, no caso) com o SKU e o número de vezes que ele foi vendido
em jan/19.

In [11]:
top_skus_count = [
    ('UTSO140 Z Y BR 0400X0470 AB 00150 02 089', 406),
    ('VEVH720 Z Y BR 0410X0800 AB 00500 02 099', 100),
    ('SLUL630 Z Y BR 0400X1130 AB 00250 02 089', 25),
    ('VEVP760 Z Y BR 0320X0880 AB 00350 02 099', 103),
    ('RTBC830 Z Y BR 0480X0550 AB 00250 10 089', 6),
    ('KVNL330 Z Y BR 0500X1040 AB 00300 05 089', 6),
    ('KVNM600 Z Y BR 0210X0770 AB 00200 05 089', 350),
    ('SLUI380 Z Y BR 0460X0970 AB 00450 10 089', 635),
    ('KVNW920 Z Y BR 0120X0450 AB 00400 02 089', 97),
    ('VEVG170 Z Y BR 0310X0900 AB 00250 05 099', 308),
    ('SLUY170 Z Y BR 0280X0870 AB 00500 02 099', 192),
    ('VEVV140 Z Y BR 0330X0630 AB 00150 02 099', 362),
    ('SLUL810 Z Y BR 0110X0720 AB 00200 05 089', 222),
    ('VEVQ340 Z Y BR 0490X0550 AB 00300 10 099', 354),
    ('SLUB390 Z Y BR 0420X0530 AB 00500 02 099', 419),
    ('VEVN770 Z Y BR 0150X0270 AB 00200 02 099', 169),
    ('WTEQ760 Z Y BR 0230X0230 AB 00450 02 089', 15),
    ('KVNL210 Z Y BR 0120X1180 AB 00200 02 089', 251),
    ('VEVO820 Z Y BR 0470X0790 AB 00450 02 099', 154),
    ('SLUU930 Z Y BR 0320X1090 AB 00250 02 089', 446),
    ('SLUL970 Z Y BR 0450X0770 AB 00150 02 089', 1519),
    ('VEVT190 Z Y BR 0340X0660 AB 00450 05 099', 256),
    ('SLUL400 Z Y BR 0270X0910 AB 00350 02 099', 122),
    ('VEVY680 Z Y BR 0470X1140 AB 00150 02 099', 315),
    ('SLUM990 Z Y BR 0220X0690 AB 00350 02 099', 178),
    ('KNIJ590 Z Y BR 0300X0780 AB 00450 02 089', 291),
    ('KVNH160 Z Y BR 0450X0610 AB 00300 02 089', 41),
    ('VEVK200 Z Y BR 0260X1050 AB 00300 02 099', 506),
    ('KVNC610 Z Y BR 0120X0690 AB 00200 02 089', 80),
    ('SLUQ150 Z Y BR 0330X0980 AB 00200 05 089', 318),
    ('XTHL520 Z Y BR 0330X0820 AB 00150 02 089', 311),
    ('VEVT190 Z Y BR 0170X1080 AB 00150 02 099', 149),
    ('KVNX880 Z Y BR 0190X0840 AB 00250 02 089', 180),
    ('VEVO290 Z Y BR 0400X1170 AB 00150 02 099', 26),
]

In [12]:
# Seu código aqui
# Solução 1
vendas_jan = 0

for sku, venda in top_skus_count: #percorrendo simultâneamente os elementos das tuplas contidas na lista, onde o primeiro elemento contem o sku e o segundo as vendas
    if sku[:3] in ['KVN', 'UTS', 'XTH']:
        vendas_jan += venda

print(vendas_jan)

1722


In [13]:
# Solução 2
vendas_jan = 0
familias = ['KVN', 'UTS', 'XTH']

for sku in top_skus_count: #percorrendo cada elemento da minha lista, que são tuplas
    if sku[0][:3] in familias: #sku[0][:3] a primeira chave acessa o primeiro elemento da tupla([0]), a segunda chave ([:3]) seleciona apenas os 3 primeiros caracteres
        vendas_jan += sku[1] #sku[1] acessa o segundo elemento da tupla que corresponde a venda

print(vendas_jan)

1722


A quantidade somada de vendas para as famílias `KVN`, `UTS` e `XTH` foram vendidas em jan/19 foi:

- A) 1961
- B) 1952
- C) 1722
- D) 2243
- E) 2897

---

## Quiz Q.2.2

Você usará a lista `vendedores` abaixo, contendo os vendedores da empresa
fictícia, e o dicionário `dict_letras`, que relaciona uma letra a algum
vendedor dentre os melhores.

In [1]:
# Aqui está a lista de vendedores, incluindo os melhores vendedores
vendedores = [
    4549564635, 4549625865, 4549659011, 4549068724, 4549212700, 4549991918, 4549153288, 4549331363,
    4549360321, 4549780924, 4549298519, 4549327159, 4549474538, 4549474203, 4549973610, 4549362906,
    4549213714, 4549061782, 4549225977, 4549535929, 4549773845, 4549341554, 4549676015, 4549967280,
    4549120227, 4549437819, 4549136351, 4549595669, 4549513928, 4549706613, 4549519614, 4549602217,
    4549957764, 4549085263, 4549297426, 4549761652, 4549331606, 4549907400, 4549099259, 4549057934,
    4549923798, 4549452748, 4549513556, 4549439333, 4549919150, 4549843722, 4549034545, 4549401011,
    4549965150, 4549695318, 4549535864, 4549676139, 4549213950, 4549613640, 4549247865, 4549568924,
    4549209648, 4549801787, 4549053964, 4549253513, 4549022115, 4549689487, 4549999591, 4549589383,
    4549338203, 4549867184, 4549244990, 4549407495, 4549410321, 4549565634, 4549255762, 4549729277,
    4549762448, 4549238354, 4549914723, 4549463488, 4549228492, 4549782058, 4549174417, 4549550542,
    4549241217, 4549783305, 4549471986, 4549317071, 4549227401, 4549408051, 4549701258, 4549396791,
    4549582356, 4549793542, 4549772849, 4549417927, 4549718202, 4549462229, 4549533615, 4549946486,
    4549147011, 4549732274, 4549994672, 4549723655, 4549518481, 4549343224, 4549212374, 4549743130,
    4549728884, 4549078871, 4549592498, 4549195967, 4549654038, 4549555453, 4549630148, 4549831522,
    4549398333, 4549326394, 4549612447, 4549836639, 4549093299, 4549511854, 4549126665, 4549139818,
    4549127965, 4549657551, 4549081078, 4549260950, 4549479537, 4549155973, 4549046890, 4549254616,
    4549810753, 4549692175, 4549614159, 4549104887, 4549564415, 4549804397, 4549092110, 4549668139,
    4549453584, 4549279132, 4549976545, 4549798990, 4549799313, 4549033990, 4549625591, 4549546331,
    4549996249, 4549090844, 4549767841, 4549326150, 4549239682, 4549600189, 4549520325, 4549263076,
    4549424602, 4549920845, 4549662610, 4549655984, 4549934896, 4549193354, 4549799442, 4549488139,
    4549666905, 4549849416, 4549981797, 4549091846, 4549182025, 4549422536, 4549403704, 4549344697,
    4549496343, 4549713225, 4549948743, 4549585545, 4549894395, 4549467060, 4549341529, 4549073643,
    4549114432, 4549882665, 4549887397, 4549108894, 4549675786, 4549924866, 4549420918, 4549227971,
    4549326230, 4549440746, 4549159270, 4549098848, 4549600412, 4549972405, 4549485225, 4549090389,
    4549073814, 4549934909, 4549323124, 4549370302, 4549570579, 4549329854, 4549481386, 4549074269,
    4549347366, 4549652195, 4549622586, 4549363969, 4549277249, 4549631073, 4549544217, 4549614386,
    4549535032, 4549532116, 4549239201, 4549711321, 4549088711, 4549034552, 4549188042, 4549727996,
    4549364368, 4549472470, 4549048384, 4549098316, 4549171303, 4549193022, 4549295683, 4549660515,
]

# E aqui o dicionário de letras. Apenas os melhores vendedores possuem caracter associado
dict_letras = {
    4549401011: 'e',
    4549532116: 'r',
    4549957764: 't',
    4549728884: ' ',
    4549471986: 't',
    4549727996: 'n',
    4549092110: ' ',
    4549801787: 'q',
    4549564635: 'Q',
    4549612447: 'i',
    4549546331: 'i',
    4549114432: 't',
    4549585545: 'n',
    4549973610: 'n',
    4549452748: 's',
    4549463488: 'e',
    4549098316: 't',
    4549422536: 'e',
    4549555453: 'c',
    4549600189: 'l',
    4549924866: 'o',
    4549174417: 'n',
    4549676139: ' ',
    4549341554: 'e',
    4549331363: 'i',
    4549159270: ' ',
    4549772849: 'a',
    4549407495: 'a',
    4549462229: ' ',
    4549662610: ' ',
    4549068724: 'u',
    4549439333: ' ',
    4549481386: ' ',
    4549622586: 't',
    4549193022: 'a',
    4549595669: 'n',
    4549057934: 'o',
    4549046890: 'c',
    4549279132: 'm',
    4549723655: 'e',
    4549213714: 'h',
    4549255762: 'r',
    4549034552: 'i',
    4549999591: 'u',
    4549692175: 'o',
    4549849416: 'c',
    4549126665: 'n',
    4549934909: 'e',
}

In [14]:
# Seu código aqui
my_string = ''
for v in vendedores:
  if v in dict_letras.keys():
    my_string += dict_letras[v]
print(my_string)

Quinhentos e quarenta e cinco mil cento e trinta


Temos uma lista vendedores acima, contendo os vendedores da empresa fictícia.
Entre eles, estão vendedores que tiveram melhor desempenho nos últimos meses.
Para cada um deles, há um caractere associado no dicionário dict_letras.

Para encontrar a resposta, você precisa iterar na lista
e criar uma string a partir dos caracteres dos melhores vendedores.

Isso significa: iterar em vendedores e se o item estiver contido entre as chaves do dicionário,
você adiciona o caractere correspondente à string da mensagem final.

- A) 456285
- B) 464140
- C) 545130
- D) 563245
- E) 326478

---